### PR TITLE
[fmha_v2] perf: use non-interleaved paged KV input for trtllm sm90/120 prefill

### PR DESCRIPTION
**Pull request by akhilg-nv**
## Motivation

Dependent on https://github.com/sgl-project/sglang/pull/23112 and https://github.com/flashinfer-ai/flashinfer/pull/4124. Improves TRT-LLM SM90/120 prefill performance to beat FA3 by removing an expensive copy-operation that was previously needed to pass interleaved KV to the kernel API.

## Modifications

Allocate KV together (no functional change to any other backend), allowing us to use the new indexing mode to read data efficiently for fmha_v2.

> the below accuracy/perf data was collected and summarized by Claude

## Accuracy Tests

PQA Diamond, 198 questions x 8 repeats (1,584 rollouts per config), temperature 0.6,
top-k 20, max-tokens 81920.

**Qwen3-Next-80B-A3B-Thinking-FP8:**

| Config | KV dtype | Score | Repeat range |
|---|---|---|---|
| FA3 | bf16 | 0.725 | 0.712-0.742 |
| fmha_v2 (trtllm_mha) | bf16 | 0.723 | 0.702-0.753 |
| FA3 + MTP | bf16 | 0.734 | 0.712-0.758 |
| fmha_v2 prefill + MTP (fa3 verify) | bf16 | 0.721 | 0.707-0.732 |
| FA3 | fp8_e4m3 | 0.727 | 0.707-0.747 |
| **fmha_v2 (trtllm_mha)** | **fp8_e4m3** | **0.724** | 0.682-0.753 |

Per-repeat std is ~0.015, so every row is a tie. The fp8 fmha_v2 row ran the
**zero-copy path in this PR**, making it a 1,584-rollout validation of the new KV
addressing against the e4m3 kernel.

**Qwen3-32B-FP8, fp8 KV:**

| Config | Score | Range |
|---|---|---|
| FA3 | 0.630 | 0.606-0.677 |
| fmha_v2 | **0.630** | 0.606-0.657 |

Exact tie.

**MTP acceptance** (draft/target property -- matching rates show fmha_v2 prefill does
not perturb the target distribution):

| Config | Accept len | Accept rate |
|---|---|---|
| FA3 + MTP (bf16) | 2.37 | 0.46 |
| fmha_v2 prefill + MTP (bf16) | 2.45 | 0.48 |

<details>
<summary>Repro</summary>

```bash
python3 -m sglang.test.run_eval --port 30000 --eval-name gpqa \
  --num-examples 198 --repeat 8 --num-threads 16 \
  --max-tokens 81920 --thinking-mode qwen-3 --temperature 0.6 --top-k 20
```

</details>

## Speed Tests and Profiling

**Before/after on identical config** (Qwen3-Next-80B-A3B-Thinking-FP8, bf16 KV,
`--chunked-prefill-size 2048`). "Before" is the original `torch.stack([k_cache, v_cache], dim=1)`
full-pool copy; "after" is this PR's non-interleaved zero-copy path.

| ISL | metric | before (stack) | after (zero-copy) | FA3 |
|---|---|---|---|---|
| 1k | total tok/s | 11017 | **14940** | 14833 |
| 1k | median TTFT (ms) | 1696 | **719** | 841 |
| 8k | input tok/s | 13853 | **22266** | 22982 |
| 8k | median TTFT (ms) | 6566 | **2952** | 2713 |
| 128k | input tok/s | 14859 | **25506** | 29363 |
| 128k | median TTFT (ms) | 11215 | **5706** | 4764 |

**Root cause of the "before" column** (torch profiler, 2.7k-token request,
24 prefill / 384 decode layer-calls):

| | fmha_v2 (stack) | FA3 |
|---|---|---|
| prefill attention | 1.59 ms | 0.98 ms |
| decode attention | 15.4 ms (bf16 XQA) | 6.8 ms |
| KV-pool stack copy | **120.5 ms** (`direct_copy_kernel`, 1008 calls) | -- |

The stack copied the entire per-layer KV pool (~4 GB, ~5 ms) per prefill layer call,
per chunk, regardless of batch size. Causality check -- only the pool size varies:

| mem-fraction-static | KV pool per side | input tok/s @8k | median TTFT (ms) |
|---|---|---|---|
| 0.8 | 24.0 GB | 13853 | 6566 |
| 0.55 | 15.0 GB | 19292 | 3809 |

**At the default chunk size** (fp8 KV, `--chunked-prefill-size 8192`; code pinned via
git worktrees, sglang@3c02143bb7 + flashinfer fmha_v2_test = perf_qtile@2a14301a +
cherry-picked 49b01545). Cells: input tok/s / median TTFT ms / median TPOT ms.

| ISL | FA3 | fmha_v2 zero-copy | ratio |
|---|---|---|---|
| 64k | 40368 / 2459 / 15.3 | **40918 / 2581 / 14.85** | 1.01 |
| 128k | 38643 / 3236 / 13.9 | **39309 / 3299 / 13.10** | 1.02 |

Confirmation trace (`wt_fmhav2_fp8_64k`): zero pool-sized `aten::cat` ops, GPU busy
1.38 s (FA3: 1.39 s), fmha_v2 kernel 81.6 ms.

**Dense model** (Qwen3-32B-FP8, 64 layers / 8 KV heads / head_dim 128, TP=4 -> 16q:2kv
per rank, YaRN x4 for 128k, chunk 8192, mem 0.85). Cells: input tok/s / TTFT ms / TPOT ms.

| ISL | FA3 | fmha_v2 |
|---|---|---|
| 8k | 21992 / 3473 / 16.5 | **23113 / 3471 / 15.4** (+5.1%) |
| 32k | 18661 / 5169 / 17.3 | **19324 / 4955 / 16.8** (+3.6%) |
| 128k | 12462 / 13869 / 27.7 | **13568 / 12662 / 25.4** (+8.9%) |

Full fp8 ISL sweep at chunk 2048 (1k-128k) is in the linked results doc; parity within
~1.5% for 1k-32k, -6% at 64k and -11% at 128k at that non-default chunk size only.

<details>
<summary>Repro</summary>

```bash
export PYTHONPATH=/SGL/sglang/python:/SGL/flashinfer
export FLASHINFER_DISABLE_VERSION_CHECK=1

python3 -m sglang.launch_server --model <MODEL> --tp 4 --mem-fraction-static 0.8 \
  --disable-radix-cache --mamba-ssm-dtype float32 --moe-runner-backend deep_gemm \
  --attention-backend trtllm_mha --trust-remote-code --port 30000
# fp8 KV: add --kv-cache-dtype fp8_e4m3

python3 -m sglang.bench_serving --backend sglang --host 127.0.0.1 --port 30000 \
  --model <MODEL> --dataset-name random --random-input-len <ISL> \
  --random-output-len <OSL> --random-range-ratio 1.0 \
  --num-prompts <N> --max-concurrency <C> --warmup-requests 2
```
Per-ISL (N, C, OSL): 1k=(200,64,256), 4k=(96,32,512), 8k=(64,32,512), 16k=(32,16,512),
32k=(16,8,512), 64k=(12,4,256), 128k=(8,2,256). The 64k/128k rows are latency-bound at
low concurrency (C=4/2), not throughput-saturated.

</details>

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #32542951602](https://github.com/sgl-project/sglang/actions/runs/32542951602)
Latest PR Test (Extra): :x: [Run #32542951489](https://github.com/sgl-project/sglang/actions/runs/32542951489)
Latest PR Test (AMD ROCm 7.2): :x: [Run #32542951576](https://github.com/sgl-project/sglang/actions/runs/32542951576)

### CODE REVIEW

**Comment by nvpohanh** [comment]
@akhilg-nv could we fuse this PR into #23112 so that we only need to review one PR? thanks!

**Comment by akhilg-nv** [comment]
> @akhilg-nv could we fuse this PR into #23112 so that we only need to review one PR? thanks!

iirc we discussed offline that since it's dependent on the flashinfer change (which was merged recently) it needs to be separate. The base PR is in now, I will re-base and recheck the accuracy and perf before pinging for review again.

**Comment by nvpohanh** [comment]
[by Codex] The lint failure is caused by this PR. The trailing-whitespace hook modifies python/sglang/srt/layers/attention/trtllm_mha_backend.py at the blank line immediately before _fmha_v2_paged_inputs (around line 282). Please remove the spaces and rerun pre-commit.

Failed job: https://github.com/sgl-project/sglang/actions/runs/32542951328/job/96956296045

**Comment by YAMY1234** [comment]
@akhilg-nv Could you address the comment from Po-Han and resolve the conflict? Thanks!
